### PR TITLE
Fast exits

### LINKED ISSUE #4: support fast exits
## Scope
- add fast exit market maker to burner wallet
- do not expect client to be online (keep receipt on backend)
- integrate exit with [sunDAI <-> DAI market](https://github.com/leapdao/burner-wallet/issues/5)

### Publicly visible delivery
- documentation in repo

## Deliverables
- working button on burner wallet

## Roles: 
bounty gardener: name / share 
bounty worker: name / share 
bounty reviewer: name / share

### PR DESCRIPTION
**Pull request by troggy**
Resolves #4

Requires: https://github.com/leapdao/leap-core/pull/85

### CODE REVIEW

File: src/components/Exchange.js [hunk 1]
```diff
@@ -120,7 +139,40 @@ export default class Exchange extends React.Component {
...
+    xdaiweb3.getColor(tokenAddr)
+    .then(color => {
+      return fetch(
+      `https://2nuxsb25he.execute-api.eu-west-1.amazonaws.com/testnet/exits/${account}/${color}`,
+      { method: "GET", mode: "cors" }
+      );
+    })
+    .then(response => response.json())
+    .then(rsp => {
+      console.log(rsp);
```
Comment (TimDaub) [inline src/components/Exchange.js]:
```suggestion
```

File: src/components/Exchange.js [hunk 2]
```diff
@@ -1456,25 +1496,88 @@ export default class Exchange extends React.Component {
...
+                          headers: {
+                              "Content-Type": "application/json",
+                          },
+                          body: JSON.stringify(setup),
+                        }
+                      );
                     })
+                    .then(response => response.json())
+                    .then(rsp => {
+                      console.log(rsp);
```
Comment (TimDaub) [inline src/components/Exchange.js]: This closes the exchange dialogue after successfully exiting.

```suggestion
                      this.setState({amount:"",daiToXdaiMode:false})
                      console.log(rsp);
```

File: src/components/Exchange.js [hunk 3]
```diff
@@ -1456,25 +1496,88 @@ export default class Exchange extends React.Component {
...
+                        tx: transferTx,
+                        effectiveBlock: periodBlockRange(transferTx.blockNumber)[1],
+                        sigHashBuff,
+                        sigHash,
+                      };
+
+                      return Tx.signMessageWithWeb3(this.props.web3, sigHash);
+                    })
+                    .then(sig => {
+                      const vBuff = Buffer.alloc(32);
```
Comment (TimDaub) [inline src/components/Exchange.js]: Not sure where `window.Buffer` comes from. As far as I know it needs to be polyfilled with https://github.com/feross/buffer. Maybe a depencency of the burner-wallet already includes it into the `window.` object. I'd still include it into `package.json`, just to make sure.

Comment (troggy) [reply src/components/Exchange.js]: good catch. Thanks Tim! We use it like that without properly declaring in bridge ui as well 🤦‍♂️

Comment (troggy) [reply src/components/Exchange.js]: I guess it is because of the webpack: https://github.com/feross/buffer/issues/161#issuecomment-320367138

Comment (troggy) [reply src/components/Exchange.js]: Indeed, webpack it is:
https://webpack.js.org/configuration/node/#nodebuffer
https://github.com/webpack/node-libs-browser

I'd just do nothing here

**Comment by troggy** [comment]
@TimDaub I've updated the PR to work with meta accounts. Most of the logic moved into leap-core: https://github.com/leapdao/leap-core/pull/85

**Comment by TimDaub** [comment]
> @TimDaub I've updated the PR to work with meta accounts.

Just tested it on Firefox and it worked great! LGTM

**Pull request merged**